src/hierarchies/mfa_hierarchy.cc

```diff
@@ -0,0 +1,288 @@
+#include "mfa_hierarchy.h"
+
+#include <google/protobuf/stubs/casts.h>
+
+#include <Eigen/Dense>
+#include <cassert>
+#include <iostream>
+#include <stan/math/prim/prob.hpp>
+#include <vector>
+
+#include "algorithm_state.pb.h"
+#include "hierarchy_prior.pb.h"
+#include "ls_state.pb.h"
+#include "src/utils/proto_utils.h"
+#include "src/utils/rng.h"
+
+double MFAHierarchy::like_lpdf(const Eigen::RowVectorXd& datum) const {
+  return stan::math::normal_lpdf(
+      datum, state.mu + state.Lambda * state.Lambda.transpose(),
+      state.psi.cwiseSqrt());
+}
+
+MFA::State MFAHierarchy::draw(const MFA::Hyperparams& params) {
+  auto& rng = bayesmix::Rng::Instance().get();
+  MFA::State out;
+  out.mu = params.mutilde;
+  out.psi = params.beta / (params.alpha0 + 1.);
+  out.Eta = Eigen::MatrixXd::Zero(card, params.q);
+  out.Lambda = Eigen::MatrixXd::Zero(p, params.q);
+  for (size_t j = 0; j < p; j++) {
+    out.mu[j] =
+        stan::math::normal_rng(params.mutilde[j], sqrt(params.phi), rng);
+
+    out.psi[j] = stan::math::inv_gamma_rng(params.alpha0, params.beta[j], rng);
+
+    for (size_t i = 0; i < params.q; i++) {
+      out.Lambda(j, i) = stan::math::normal_rng(0, 1, rng);
+    }
+  }
+
+  for (size_t i = 0; i < card; i++) {
+    for (size_t j = 0; j < params.q; j++) {
+      out.Eta(i, j) = stan::math::normal_rng(0, 1, rng);
+    }
+  }
+
+  return out;
+}
+
+void MFAHierarchy::initialize_state() {
+  state.mu = hypers->mutilde;
+  state.psi = hypers->beta / (hypers->alpha0 + 1.);
+  state.Eta = Eigen::MatrixXd::Zero(card, hypers->q);
+  state.Lambda = Eigen::MatrixXd::Zero(p, hypers->q);
+}
+
+void MFAHierarchy::initialize_hypers() {
+  if (prior->has_fixed_values()) {
+    // Set values
+    hypers->mutilde = bayesmix::to_eigen(prior->fixed_values().mutilde());
+    p = hypers->mutilde.size();
+    hypers->beta = bayesmix::to_eigen(prior->fixed_values().beta());
+    hypers->phi = prior->fixed_values().phi();
+    hypers->alpha0 = prior->fixed_values().alpha0();
+    hypers->q = prior->fixed_values().q();
+
+    // Check validity
+    if (p != hypers->beta.rows()) {
+      throw std::invalid_argument(
+          "Hyperparameters dimensions are not consistent");
+    }
+    for (size_t j = 0; j < p; j++) {
+      if (hypers->beta[j] <= 0) {
+        throw std::invalid_argument("Shape parameter must be > 0");
+      }
+    }
+    if (hypers->alpha0 <= 0) {
+      throw std::invalid_argument("Scale parameter must be > 0");
+    }
+    if (hypers->phi <= 0) {  // TODO check
+      throw std::invalid_argument("Diffusion parameter must be > 0");
```

Comment:
What's up with this? Also, you should change error messages from `>` to `<=`

Reply:
You are right, all these exceptions still need to be checked, we also need to work on some automatic hyperparameter initialization as proposed in the paper of this model.